rework testcontainer usage

mgmeiner opened the pull request:
When i implemented the other pull request i noticed that the tests clashed with my local running _kafka_ + _zookeeper_ and the schema-registry clashed with a completely different service running on that port (8081).
So i thought fixing this by choosing a random and non-clashing ports which is done by _testcontainers_ out of the box.

This removes the _testcontainers_ setup with docker-compose and replaces it with _testcontainer_'s container definition by code. 

For applying the random and non-clashing ports to the `EnvConfig` during startup I introduced a customizer for the config which executes when instantiating it for the _Koin_ module. I'm new to _Koin_ so there might be a better solution to this problem 🙂.

I also updated the used kafka version for tests to: _5.5.1_.

wlezzar commented:
nice! Thanks a lot for the PR : ) .  I will review this ASAP

wlezzar commented on zoe-cli/src/commands/main.kt:
```diff
@@ -182,7 +185,7 @@ fun mainModule(context: CliContext) = module {
                 LocalConfigDirUrlProvider(context.configDir)
             )
-        ).createConfig(context.env)
+        ).createConfig(context.env).let { envConfig -> customize(envConfig) }
```
Instead of injecting the customizer here, what do you think of rather overriding the `EnvConfig` bean during tests as explained in [this section of the documentation](https://doc.insert-koin.io/#/koin-core/modules?id=overriding-definition-or-module)?

We could define a testing instance of `EnvConfig` that we build from scratch in the tests and fullfill it with values from the test containers, etc. and then use it as an override.

mgmeiner replied:
Yes that's definitely a better solution! I will update it accordingly 😊

wlezzar commented on zoe-cli/test/com/adevinta/oss/zoe/cli/helpers.kt:
```diff
@@ -72,17 +76,28 @@ suspend fun ExpectScope.zoe(
 
-val testConfDir = File("testResources/env")
+private val testModule = module {
+    single { customizeEnvConfig }
```
I think you could instead override the whole `EnvConfig` by an instance created from scratch for the purpose of testing (see [my above comment](https://github.com/adevinta/zoe/pull/6/files#r456514406))

wlezzar commented on zoe-cli/test/com/adevinta/oss/zoe/cli/testcontainers.kt:
```diff
@@ -0,0 +1,57 @@
+ * Context which holds the started containers for easy access in application setup customizers.
+ */
+object TestcontainersContext {
+    private lateinit var kafkaContainer: KafkaContainer
+    val kafka get() = kafkaContainer
```
You could this:
You could do this:
```suggestion
    private lateinit var kafkaContainer: KafkaContainer
          private set
```

mgmeiner replied:
way better now! Did not think of using a simple private setter on a lateinit var.

wlezzar commented on zoe-cli/test/com/adevinta/oss/zoe/cli/testcontainers.kt:
```diff
@@ -0,0 +1,57 @@
+    private lateinit var kafkaContainer: KafkaContainer
+    val kafka get() = kafkaContainer
+
+    private lateinit var schemaRegistryContainer: SchemaRegistryContainer
+    val schemaRegistry get() = schemaRegistryContainer
```
Same here:
```suggestion
    private lateinit var schemaRegistryContainer: SchemaRegistryContainer
       private set
```

wlezzar commented on zoe-cli/test/com/adevinta/oss/zoe/cli/testcontainers.kt:
```diff
@@ -0,0 +1,57 @@
+    }
+}
+
+class SchemaRegistryContainer(kafka: KafkaContainer) :
```
can you inject only the `network` here instead of the whole `KafkaContainer` to make the intent clearer?

wlezzar commented on zoe-cli/test/com/adevinta/oss/zoe/cli/helpers.kt:
```diff
@@ -72,17 +76,28 @@ suspend fun ExpectScope.zoe(
-        waitingFor("broker", Wait.forListeningPort())
-        waitingFor("schema-registry", Wait.forHttp("/subjects").forStatusCode(200))
-    }
+class TestcontainersListener : TestListener {
```
Now that we have a `testcontainers.kt` you can move this listener there :)

mgmeiner replied:
Good point 😅!

mgmeiner commented on zoe-cli/src/commands/main.kt:
```diff
@@ -98,31 +76,27 @@ class ZoeCommandLine : CliktCommand(name = "zoe") {
             else -> Level.ALL
         }
-
-        loadKoinModules(
```
@wlezzar I moved the koin module definition out of the zoe main Command  and moved it into the main entry point of the whole cli. I think this is better place as there the whole koin lifecycle is already controlled + it is more clear in which order modules can be defined

wlezzar replied:
Hi @mgmeiner . I took a look at the PR. I see your point of moving the koin modules loading in the same place where koin is instantiated.

However, I feel this is not compatible with the philosophy of Clikt and the application: basically, as the Clikt commands are independently nested, each should be able to independently create its own beans and modules. Currently, this is achievable by making the Clikt command instantiate a koin module itself and this becomes available to the child commands (like the main CliktCommand is currently doing). If we make all the modules be loaded at the top level we will need to create more links and coupling across the commands themselves. Makes sense? what do you think?

mgmeiner replied:
@wlezzar Ok i get your point but the problem in this case is the order in which the koin modules are loaded because we need to override the EnvConfig bean and so this must happen after the main module is loaded and if this configuration logic happens during "run phase" of the command its quite hard to intercept it. 

One option  would be to pass the additional koin modules as constructor parameter or provide a configuration function on ZoeCommandLine which passes additional modules. Latter option might be more readable and clear as Clikt does it the same way. 

Like:
```
val command =
    ZoeCommandLine()
         .subcommands(
              topicsCommand(),
              schemasCommand(),
              groupsCommands(),
              configCommands(),
              lambdaCommands(),
              versionCommands()
         )
         .context(customizeContext)
         .additionalKoinModules(additionalModules)
```
Another solution could be to pass a simple callback function which executes between configuration and actual "run".
Maybe you have a better solution in mind?

wlezzar replied:
Ok I completely see your point now.

I like the callback solution: I'm thinking of the following: create an `EnvConfigCustomizer` that extends `(EnvConfig) -> EnvConfig`. Then [in this block](https://github.com/adevinta/zoe/blob/master/zoe-cli/src/commands/main.kt#L180-L185) where we load the `EnvConfig`, we can check if such a bean exist in the module. If it does exist, it gets executed as the final step there. This `EnvConfigCustomizer` will be injected in the tests to override the configs.
What do you think of this a solution?

I also agree that there is something to clarify / fix in the koin modules loading but I suggest tackling this in a different PR / issue. What do you think?

(ps. sorry for the delay, I'm fully back now)

mgmeiner commented on zoe-cli/src/commands/main.kt:
```diff
@@ -174,135 +148,6 @@ class ZoeCommandLine : CliktCommand(name = "zoe") {
 
-@FlowPreview
-@ExperimentalCoroutinesApi
-fun mainModule(context: CliContext) = module {
```
The whole module definition is also moved to the main entry point.

mgmeiner commented on zoe-cli/src/main.kt:
```diff
@@ -40,6 +64,10 @@ fun <T> withZoe(customizeContext: Context.Builder.() -> Unit = {}, action: (ZoeC
                 )
                 .context(customizeContext)
 
+        val modules = listOf(mainModule { command.buildCliContext() }) + additionalModules
```
here it is more clear now in which order the koin modules are defined and we are able to pass additional modules. In this particular case: the test module which overrides the EnvConfig definition

wlezzar commented:
@mgmeiner nice! : ) . I will take a look at these chabnes today (sorry I have been off these last couple of days). I will give you some feedback later today 👍

mgmeiner commented:
I reverted everything to my first solution which customizes the `EnvConfig` with a simple kotlin function `(EnvConfig) -> EnvConfig` and refactored it slightly. 

The private `customize` function looks for beans/functions, which implement this function-pattern: `(T) -> T`, and if one or more exists it invokes them + reassigns the result as the actual bean.  
I skipped creating a dedicated interface for the customizer but I'm not sure if its better to have one or to simply use the kotlin function definition as it is now. This solution now would basically work for every bean without the need for custom 'Customizer'-Interfaces but might not be obvious at first sight.

I hope this is what you're looking for 😊.

wlezzar reviewed (approved):
Looks super good to me, thanks a lot @mgmeiner !

wlezzar commented:
I'm ready to merge it if you are done @mgmeiner ?

mgmeiner commented:
@wlezzar Yeah :) But one more thing: We could speedup tests by not starting and stopping the Containers for every Spec and instead just start the Containers once (with `init` on the Testcontainers Context object). The teardown is handled automatically by Testcontainers. If you like i could add that too unless you want a fresh kafka instance per Spec.

wlezzar commented:
> @wlezzar Yeah :) But one more thing: We could speedup tests by not starting and stopping the Containers for every Spec and instead just start the Containers once (with `init` on the Testcontainers Context object). The teardown is handled automatically by Testcontainers. If you like i could add that too unless you want a fresh kafka instance per Spec.

I actually had this in my todo list, so yes completely agree ^^ . The tests are designed to not impact each others, so one cluster for all the tests would be enough. So if you would like to do it in this PR, would be super cool : )

mgmeiner commented:
Ok i squashed the previous commits and added another one for the "starting-only-once".
Thanks for your review and support on this PR 😊

wlezzar commented:
Thanks a lot @mgmeiner : ) . And sorry it took so much time..
